elizaOS/eliza
# feat: Separate Knowledge system + Multi-Agent RAG Optimization

# Relates to:


# Risks

Low Risk, as the new RAG system is completely separate, and must be enabled by the user to use as of now.

# Background

## What does this PR do?

## What kind of change is this?

Improvements (misc. changes to existing features)

This improves the knowledge system by separating it from the memories table, while allowing users to better manager their knowledge and have true separation of actions.

Additionally, allowing multiple agents to "share" knowledge to avoid duplicating entries for better optimizations.

All database adapters were updated to have the new functions to create/get/search/delete knowledge.

## Why are we doing this? Any context or related work?

This change is a critical change as it helps the agent scale, reduce database overlap, and separates the agents actions, to ensure proper segregation.

Additionally, this also allows users to create a knowledge folder of txt/md files and reference them in their character knowledge, without having to paste the full contents in character file.

# Documentation changes needed?

My changes require a change to the project documentation.

Minimal changes to exiting knowledge docs, to outline how to enable the new rag knowledge system.


# Testing

## Where should a reviewer start?

## Detailed testing steps

- Enable the new rag knowledge 
![image](https://github.com/user-attachments/assets/4e1ca71f-800c-47f3-9bd4-6c4103ea2aa5)

- Create a "knowledge" folder within your characters folder
![image](https://github.com/user-attachments/assets/62b6bfd3-29ad-4a55-81fc-78fcd7d26f6e)

- Add files in the root of knowledge, or create additional subfolders for organization

- Set your knowledge items in your character file directly, or objects specifying the path, and if it should be a shared item or not, depending on multi characters or not.
![image](https://github.com/user-attachments/assets/5602960c-2aef-4085-82a9-e13624acf1b6)

- Start agent(s) as normal, and watch knowledge get processed!

## Screenshots
### Before
### After

Knowledge table with information and split into chunks, along with main document, shared documents, or agent specific documents in the database.
![image](https://github.com/user-attachments/assets/ce69f38e-7fe0-4772-9137-1f6dae424006)

# Deploy Notes
## Database changes
 - New knowledge table is created
 -

## Deployment instructions

## Discord username
ninja_dev
